### PR TITLE
fix: optimize attention decomposition

### PR DESCRIPTION
**Pull request by fs-eire**
# Description

Align attention decomposition behavior with the ONNX graph.

This PR does the following:

- Restrict attention-mask Select choices to scalars while preserving mask polarity and -inf semantics.
- Preserve static arange operations as TensorRT LINSPACE Fill layers.
- Add regression coverage for attention masks and static arange conversion.

With this change, Torch-TensorRT is able to generate optimized engine with a pattern that an attention using a causal mask AND a padding mask, when `decompose_attention=True`.

Fixes # (issue)

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 1]
```diff
@@ -485,18 +485,21 @@ def scaled_dot_product_attention_decomposition(
...
     device = query.device
 
-    if is_causal or attn_mask is not None:
-        attn_bias = torch.zeros((L, S), dtype=query.dtype, device=device)
-
     if is_causal:
         assert attn_mask is None, "attn_mask must be None when is_causal=True"
+        attn_bias = torch.zeros((L, S), dtype=query.dtype, device=device)
         temp_mask = torch.ones((L, S), dtype=torch.bool, device=device).tril(diagonal=0)
         attn_bias = attn_bias.masked_fill(temp_mask.logical_not(), float("-inf"))
```
Comment (micwill755) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:492]: For the bool attn_mask path we switched to scalar where(0 / -inf), but is_causal still uses masked_fill. Should the causal path use the same scalar where pattern?

File: py/torch_tensorrt/dynamo/conversion/impl/arange.py [hunk 2]
```diff
@@ -68,9 +68,20 @@ def arange(
...
+        step_tensor = get_trt_tensor(
+            ctx, step, name + "_step", dtype=trt.int32, min_rank=1
+        )
+        fill_layer = ctx.net.add_fill(
+            output_shape, trt.FillOperation.LINSPACE, trt.int32
+        )
+        fill_layer.set_input(1, start_tensor)
+        fill_layer.set_input(2, step_tensor)
+        set_layer_name(fill_layer, target, f"{name}_arange_fill", source_ir)
+        return fill_layer.get_output(0)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/arange.py:135]: Good finding! Is this arange op upstream of attention? If so, it should resolve the perf issue for both IAttention path and Decomposed path, right?

Besides, I have an addition to arange converter. Can you review https://github.com/pytorch/TensorRT/pull/4456?

Comment (fs-eire) [reply py/torch_tensorrt/dynamo/conversion/impl/arange.py:135]: > Is this arange op upstream of attention? If so, it should resolve the perf issue for both IAttention path and Decomposed path, right?

yes

> #4456

The change should be good. I think you can merge it first and then I can combine my change based on your change.
